### PR TITLE
Add Sanction List

### PR DESCRIPTION
**Pull request by ppoliani**
Add sanction list based to the discussion here https://github.com/sigma0-xyz/zkbitcoin/issues/5.

The logic goes like this:

1. Start a thread that periodically fetches the latest list of sanctioned BTC addresses from [here](https://www.treasury.gov/ofac/downloads/sanctions/1.0/sdn_advanced.xml)
2. If the list is fresh it will extract all BTC addresses from that XML file and store in-memory
3. Exposes a `is_sanctioned` method that returns true if the address is in the sanction list.

### CODE REVIEW

**Comment by ppoliani** [comment]
It takes around 35s to update fetch the latest OFAC list and extract all addresses from XML to the in memory hashmap. The logic runs in a separate non-blocking thread so it should not cause any issues. Also, they update the list every few days so it's gonna run that often

**Comment by ppoliani** [comment]
I believe we should start this thread when we run `StartCommitteeNode`. It should wait for the fist sync of the OFAC list before the node itself starts.

File: src/address_verifier.rs [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+        let value = re.find(&str_value)?.context("no regex result")?.as_str();
+
+        Ok(value.parse()?)
+    }
+
+    /// read the first few bytes from the remote XML file and extract the last update date.
+    /// If there is no fresh data we can skip the parsing of XML which is slow.
+    async fn publish_date() -> Result<i64> {
+        let res =
+            reqwest::get("https://www.treasury.gov/ofac/downloads/sanctions/1.0/sdn_advanced.xml")
```
Comment (ppoliani) [inline src/address_verifier.rs:40]: move this to constant value

File: src/address_verifier.rs [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+
+        Ok(date)
+    }
+
+    /// Runs the OFAC list syncronization. Downloads the remote XML file and extracts the sanctioned addresses
+    async fn sync(sanctioned_addresses: Arc<RwLock<HashMap<String, bool>>>) -> Result<()> {
+        info!("OFAC list syncing...");
+        let start = Instant::now();
+        let res =
+            reqwest::get("https://www.treasury.gov/ofac/downloads/sanctions/1.0/sdn_advanced.xml")
```
Comment (ppoliani) [inline src/address_verifier.rs:69]: same here

File: src/address_verifier.rs [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+                Err(e) => {
+                    error!("Error parsing xml: {e}");
+                    break;
+                }
+                _ => {}
+            }
+        }
+
+        let duration = start.elapsed();
+        info!("OFAC synced in {:?}", duration);
```
Comment (ppoliani) [inline src/address_verifier.rs:113]: use `debug!` here

File: src/address_verifier.rs [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+
+            loop {
+                interval.tick().await;
+
+                let Ok(publish_date) = Self::publish_date().await else {
+                    error!("couldn't extract the OFAC document publish date");
+                    continue;
+                };
+
+                info!("{:?}", *last_update.read().await);
```
Comment (ppoliani) [inline src/address_verifier.rs:135]: remove this one

File: src/address_verifier.rs [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+                let mut last_update = last_update.write().await;
+                *last_update = publish_date;
+
+                if let Err(error) = Self::sync(Arc::clone(&sanctioned_addresses)).await {
+                  error!("OFAC list sync error: {}", error);
+                };
+            }
+        })
+        .await
+        .unwrap();
```
Comment (ppoliani) [inline src/address_verifier.rs:150]: avoid unwrapping here

**Comment by mimoo** [comment]
BTW let me know when this is ready for review (looks like CI doesn't pass atm)

**Comment by ppoliani** [comment]
> BTW let me know when this is ready for review (looks like CI doesn't pass atm)

Fixed the lint issues. You can review this :)

File: src/committee/orchestrator.rs [hunk 6]
```diff
@@ -268,24 +269,30 @@ pub struct Orchestrator {
...
+            compliance,
         }
     }
 
     /// Handles bob request from A to Z.
     pub async fn handle_request(&self, bob_request: &BobRequest) -> Result<BobResponse> {
         // Validate transaction before forwarding it, and get smart contract
+        bob_request
+            .check_compliance(Arc::clone(&self.compliance))
+            .await?;
```
Comment (mimoo) [inline src/committee/orchestrator.rs:295]: wondering: should individual nodes perform that check as well? What would be the rational for not performing that check at the node level (and would node be liable if something bad happens there...)

File: src/committee/orchestrator.rs [hunk 7]
```diff
@@ -567,15 +574,26 @@ pub async fn run_server(
     let address = address.unwrap_or("127.0.0.1:6666");
     info!("- starting orchestrator at address http://{address}");
 
+    let mut compliance = Compliance::new();
+    // Orchestrator should sync the Sanction ist before doing anything else
```
Comment (mimoo) [inline src/committee/orchestrator.rs:578]: sanction list*

File: src/compliance.rs [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+pub struct Compliance {
+    sanctioned_addresses: Arc<RwLock<HashMap<String, bool>>>,
+    last_update: Arc<RwLock<i64>>,
+}
+
+impl Default for Compliance {
+    fn default() -> Self {
+        Self::new()
+    }
+}
```
Comment (mimoo) [inline src/compliance.rs:24]: weird :D

File: src/compliance.rs [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+
+impl Compliance {
+    const BTC_ID: &'static str = "344";
```
Comment (mimoo) [inline src/compliance.rs:27]: can you document where you got that value from? (if it's possible)

File: src/compliance.rs [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+    ) -> Result<()> {
+        let publish_date = Self::publish_date().await?;
+
+        if *last_update.read().await >= publish_date {
+            info!("Sanction list is up-to-date");
+            return Ok(());
+        }
+
+        let mut last_update = last_update.write().await;
+        *last_update = publish_date;
```
Comment (mimoo) [inline src/compliance.rs:94]: we should do this last in case something messes up below

File: src/compliance.rs [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+        Ok(())
+    }
+
+    /// Periodically fetces the latest list from OFAC_URL and updates the local list
+    pub fn start(&self) -> JoinHandle<()> {
+        let sanctioned_addresses = Arc::clone(&self.sanctioned_addresses);
+        let last_update = Arc::clone(&self.last_update);
+
+        spawn(async move {
+            let mut interval = interval(Duration::from_secs(600));
```
Comment (mimoo) [inline src/compliance.rs:152]: we should hardcode that value somewhere

File: src/committee/orchestrator.rs [hunk 12]
```diff
@@ -567,15 +574,26 @@ pub async fn run_server(
...
         pubkey_package,
         committee_cfg,
-        member_status: member_status_state,
-    };
+        member_status_state,
+        Arc::clone(&compliance),
+    );
+
+    // Sync sanction list in a parallel thread
+    compliance.start();
```
Comment (mimoo) [inline src/committee/orchestrator.rs:596]: don't you have to use the joinhandle returned?

**Review by mimoo — APPROVED** [review]
This is amazing :) thanks so much again for these PRs!

One thing that'd be great to have is a test just to easily see what the sanction list is, how big it is, and how much time it takes to sync

File: src/committee/orchestrator.rs (hunk 12 above)
Comment (ppoliani) [reply src/committee/orchestrator.rs:596]: not if you don't want to wait for it to complete

File: src/compliance.rs (hunk 11 above)
Comment (ppoliani) [reply src/compliance.rs:152]: yes I will do that

**Comment by ppoliani** [comment]
> Sanction ist

I've added a comment regarding this.

https://github.com/sigma0-xyz/zkbitcoin/pull/30#issuecomment-1915156179

It's an XML file and is 92MB big.

File: src/compliance.rs (hunk 8 above)
Comment (ppoliani) [reply src/compliance.rs:24]: oops. I blindly followed clippy suggestion. Will change this to Derive macro

File: src/compliance.rs (hunk 9 above)
Comment (ppoliani) [reply src/compliance.rs:27]: Oh I couldn't find any documentation. This is a gov document, didn't expect anything better to be honest. I had to look into the XML file and find that number.

File: src/committee/orchestrator.rs (hunk 6 above)
Comment (ppoliani) [reply src/committee/orchestrator.rs:295]: We could, but we've got orchestrator as a centralized entity that coordinates the entire process.

**Pull request merged**